### ISSUE TITLE
Merge redesign of ocaml.org into production site

### ISSUE THREAD

**Issue by amirmc** [ocaml/v2.ocaml.org#184]
The following checklist will track the progress of merging the site redesign into the main website.  Items will be added here as we work on the branch we're merging into the main repo [1].

Note that this phase of the redesign only covers the styles, layouts and architecture of the site.  There are two other elements which will be dealt with later (specifically, packages and documentation)
#### Review phase
- [x] Create new branch (done: [1])
- [x] Check over bash scripts
- [x] Run scripts in new branch (to generate, clean and move markdown pages)
- [x] Copy over remaining issues from sandbox to this repo
- [x] Manually check that pages, urls are in correct place
- [x] Run site through link validator (done via script)
- [x] Manually edit internal links to be correct
- [x] Properly link any orphaned pages (manual check)
- [x] Close all [blocking issues](https://github.com/ocaml/ocaml.org/issues?labels=blocking%2Credesign)
#### Deployment phase
- [x] Add Google Analytics code to `analytics` branch (#242) (@agarwal)
- [x] Merge `analytics` branch into `redesign` branch (@amirmc)
- [x] Change build script on VM [2] to publish from `master` (see https://github.com/ocaml/ocaml.org/issues/184#issuecomment-28819647)
- [x] Merge `redesign` branch into `master` (do this via pull request) (@amirmc)
- [x] Change DNS on ocaml.org to point to VM (@avsm)
- [x] Delete obsolete branches: `redesign`, `publish`, `analytics`, 
- [x] Delete `gh-pages` branch once we're confident DNS changes have propagated

[1] https://github.com/ocaml/ocaml.org/tree/redesign/
[2] https://github.com/ocaml/infrastructure/wiki/OCaml.org/

**Comment by amirmc** [comment]
@Chris00 @agarwal @pw374 (cc @avsm), just drawing your attention to the 'deployment' part of the above plan.  

We'll use the current VM as the production server so we have to be careful to amend the relevant scripts before merge and point the ocaml.org DNS at the appropriate place.  This means we will no longer rely on the `gh-pages` branch and it can be removed (so we can also close #151).

Please edit the above if I've forgotten anything.

**Comment by agarwal** [comment]
We actually publish from the `publish` branch. That leaves `master` for development allowing us to review each others changes before publishing them.

**Comment by amirmc** [comment]
Got it. There's an extra step then, to make sure that the scripts that push changes from `master` to `publish` are also present and correct.

**Comment by agarwal** [comment]
So actually I changed my mind. I think we can just go with `master` and delete `publish`. Any code that should be reviewed can be either in our own fork or in a new topic branch. Given the desire for rapid turnaround time, I guess this is the model we're aiming for:
- Someone submits a pull request.
- One of us with rights to merge does so, hopefully without significant delay (but already we see that the desire to have changes go live in under 5 minutes is lost).
- The build script picks up the change and it gets published.

This can all happen on `master`. The extra indirection of first going to `master`, then `publish` would be pointless in the above workflow.

**Comment by amirmc** [comment]
@Chris00 @agarwal @pw374 @avsm 

We've only got three blocking bugs left, and the trickiest one is #282 (the others are simply awaiting a minor change and merge as far as I can tell).

I'm of the opinion that we do the merge in the morning (UK time) rather than tonight.  It's already getting late and I'm not sure of getting the DNS change done (Anil only landed this afternoon).  I propose the following:
- @agarwal creates a branch (`analytics`?) with the appropriate code and we we simply merge that into `redesign` before merging `redesign` into `master`.  That way it won't appear on preview.ocaml.org and the time difference won't cause issues for Ashish.
- @Chris00 can merge `pbm` into `redesign` when he's happy things are ok.
- @pw374 can look at #282 and figure something out (we have thoughts on what to do).
- @avsm can change the DNS in the morning at the same time I merge the branch.
- @amirmc will merge `analytics` into `redesign` and then merge `redesign` into `master`

Is that ok?

**Comment by agarwal** [comment]
This sounds fine to me. I'm working on the analytics code right now, which requires some refactoring of the build commands.

**Comment by agarwal** [comment]
I've pushed the `analytics` branch. This required changes to the build system that I'm not very happy with. It's a bit of a mess IMO. We now have a Makefile that calls gen.bash which in turn calls the Makefiles again. This circularity required me to add an option to the top make commands, which will get pushed to gen.bash, which then get converted again to something needed by the sub-make commands. Well, it's working, but we should rethink the build system.

**Comment by agarwal** [comment]
Please note, that these steps should happen in this order:
- First change script on preview.ocaml.org to run `make staging`.
- Only after that, merge `analytics` into `redesign`.

**Comment by amirmc** [comment]
I'm confused. Why 'make staging' and not 'make production'?

This VM is meant to be serving the production site. 

Best wishes,
Amir

## 

sent via mobile

**Comment by Chris00** [comment]
> merge pbm into redesign when he's happy things are ok.

Done.

**Comment by agarwal** [comment]
@amirmc I meant if you're going to do it immediately. Yes, it should be `make production` if you're going to do this at the same time as updating DNS.

**Comment by amirmc** [comment]
I'm merging `analytics` into `redesign` now

**Comment by amirmc** [comment]
done and pushed.  I know there'll be a minor merge conflict when I merge to `master` but I want to double check the DNS with Anil

**Comment by amirmc** [comment]
apparently not.  I've just pushed it to my fork to check https://github.com/amirmc/ocaml.org/

**Comment by amirmc** [comment]
Merged and pushed to `master`!  Just need to check and restart the scripts now

**Comment by agarwal** [comment]
I edited the task list with an item to delete obsolete branches. What is the `ssl` branch for? Is that also obsolete? It isn't fully merged into `master`.

**Comment by amirmc** [comment]
ocaml.org is now resolving to the new site (for me) :)

**Comment by agarwal** [comment]
Me too!

**Comment by amirmc** [comment]
Dammit. Now it's gone for me.  The joy of DNS propagation.

**Comment by Chris00** [comment]
`ssl` is a branch where we use a single script to build all kind of headlnes coming from a RSS feed.  The SSL came from the fact that I needed to enable SSL for the Google feed.  I have spoken with Xavier and I will try to develop some code to directly access Inria SSL feed so one can link directly yo the archive of the post and not to the year-month.  Some processing will be needed because this is a feed of posts, not of threads. Stay tuned!

**Comment by Chris00** [comment]
I have **rebased** the `ssl` branch on `master`.  If you checked out a copy, get rid of it!

**Comment by Chris00** [comment]
On the `ssl` branch, the "Recent email threads" of the "Community" page lead directly to the last post on the sympa server.  Try it out and tell me what you think.

**Comment by agarwal** [comment]
@amirmc Do you want me to go ahead with purging the obsolete branches?

**Comment by agarwal** [comment]
@Chris00 The published site is also showing the emails. The enhancement in the `ssl` branch appears to be that each thread title is linked directly to that thread. In the published site, all the links just go to the main mailing list page. It's nice! But is that what you meant for us to look at?

**Comment by amirmc** [comment]
@agarwal `analytics`, `redesign` and `publish` can go.  We should leave `gh-pages` for a few days until we're confident that the DNS has propagated.  Once that branch goes, anyone with stale DNS records will get a 404.

**Comment by agarwal** [comment]
Done, except for `gh-pages`.

**Comment by Chris00** [comment]
@agarwal Yes — the "technology" behind the scenes is quite different though.  If there are not objections, I'll merge it into master tomorrow.

**Comment by agarwal** [comment]
`gh-pages` branch has been deleted. That was the last item in this issue, so closing.
